### PR TITLE
Introduce helper to make map-like objects.

### PR DESCRIPTION
**Pull request by cramforce**
…using `Object.create(null)`.

And use it in a few places where it should be used.

I haven't yet migrated all usages of `Object.create(null)`.

### CODE REVIEW

File: src/types.js [hunk 1]
```diff
@@ -53,6 +53,24 @@ export function toArray(arrayLike) {
...
 
 /**
+ * Returns a map-like object.
+ * If opt_initial is omitted returns an objects with a null prototype.
+ * Otherwise takes the given object and sets its prototype to null.
+ * @param {T=} opt_initial This should typically be an object literal.
+ * @return {T}
+ * @template T
+ */
+export function map(opt_initial) {
```
Comment (tdrl) [inline src/types.js]: Is this really preferable to polyfilling the ES6 `Map`?  It looks like that's already supported in Chrome, Safari, Firefox, and mostly in IE 11.

Comment (cramforce) [inline src/types.js]: @tdrl the polyfill is pretty massive and we would need it.

File: src/types.js [hunk 2]
```diff
@@ -53,6 +53,24 @@ export function toArray(arrayLike) {
...
+  if (opt_initial) {
```
Comment (jridgewell) [inline src/types.js]: I'd rather to an `Object.assign` here.

File: src/types.js [hunk 3]
```diff
@@ -53,6 +53,24 @@ export function toArray(arrayLike) {
...
+    if (opt_initial.__proto__) {
+      opt_initial.__proto__ = null;
+    }
+    return opt_initial
+  }
+  return Object.create(null);
```
Comment (jridgewell) [inline src/types.js]: Wan to [prematurely optimize](https://github.com/google/incremental-dom/pull/259)?

File: src/types.js (hunk 2 above)
Comment (cramforce) [inline src/types.js]: I see your point, but I'd rather not copy.

File: src/types.js (hunk 3 above)
Comment (cramforce) [inline src/types.js]: LOL

File: src/types.js (hunk 2 above)
Comment (cramforce) [inline src/types.js]: K, done

File: src/types.js [hunk 4]
```diff
@@ -53,6 +53,22 @@ export function toArray(arrayLike) {
 }
 
 /**
+ * Returns a map-like object.
+ * If opt_initial is provided, copies its own properties into the
+ * newly created object.
+ * @param {T=} opt_initial This should typically be an object literal.
+ * @return {T}
```
Comment (jridgewell) [inline src/types.js]: Are these types correct?

Comment (cramforce) [inline src/types.js]: Yeah. This is as good as we can do in closure compiler. The results is:

- If you pass a literal, then the return value is the inferred record type of the literal.
- If you pass nothing, then the return value is `?` (sadly)

File: build-system/eslint-rules/no-export-side-effect.js [hunk 5]
```diff
@@ -24,7 +24,9 @@ module.exports = function(context) {
               .map(function(declarator) {
                 return declarator.init
               }).filter(function(init) {
-                return init && /(?:Call|New)Expression/.test(init.type);
+                return init && /(?:Call|New)Expression/.test(init.type)
+                    // Special case creating a map object from a literal.
+                    && init.callee.name != 'map';
```
Comment (cramforce) [inline build-system/eslint-rules/no-export-side-effect.js]: You mean if it is never used?

Comment (jridgewell) [inline build-system/eslint-rules/no-export-side-effect.js]: Yup. Closure can't even DCE an unused `var x = Object.create(null)`.

**Pull request merged**